### PR TITLE
Improve test performance by avoiding on-the-fly generation of RSA and DH keys

### PR DESCRIPTION
**Pull request by DDvO**
~Use small key sizes when generating RSA keys if possible or~ 
Where appropriate, use pre-generated RSA and DH keys for more efficient tests.

On this occasion, improve the `-keyout` option of the  `req` app as needed and 
improve the documentation of its `-new`, `-newkey`, and `-keyout` options.

~*Make various test CA certs RFC 5280 compliant w.r.t. X509 extensions.~ (moved to #13719)

### CODE REVIEW

**Comment by t8m** [comment]
I think this is a bad idea. We might eventually want to block small key sizes completely. The tests with  the 2k key size are not exceedingly slow.

19. 12. 2020 22:15, 22:15, David von Oheimb <notifications@github.com> napsal/a:

**Comment by beldmit** [comment]
I agree with Tomas.

**Comment by richsalz** [comment]
FWIW I also agree that this is a very bad idea.  In addition to the original objection, tests should be based on real-world usages.  Don't do this.

**Comment by DDvO** [comment]
All right, I'll revert using smaller RSA key sizes for tests and will instead use pre-generated RSA keys as far as possible (with reasonable effort) in order to avoid needless test ineffiencies.

**Comment by InfoHunter** [comment]
> All right, I'll reverted using smaller RSA key sizes for tests and will instead
use pre-generated RSA keys as far as possible (with reasonable effort)
in order to avoid needless test inefficiencies.

You can also consider to use multi-prime RSA to have a performance boost ;-) - but pre-generated keys are more perfect.

**Comment by t8m** [comment]
Is this still WIP?

**Comment by DDvO** [comment]
> Is this still WIP?

Yes, unfortunately I haven't found the time yet to see at which points in the tests it is worth using pre-generated RSA keys.
If anyone has a hint please let me know.

**Comment by DDvO** [comment]
> > Is this still WIP?
> 
> Yes, unfortunately I haven't found the time yet to see at which points in the tests it is worth using pre-generated RSA keys.
> If anyone has a hint please let me know.

Today I took the time to detect needless inefficient RSA and DH key generation 
and to replace it by loading pre-computed keys.

**Review by paulidale — APPROVED** [review]
Have you quantified the performance improvement this results in?

**Comment by DDvO** [comment]
> Have you quantified the performance improvement this results in?

On a 64-bit plain debug build on my machine, `time make test` yielded before the optimization:
```
real	2m43.320s
user	9m21.697s
sys	0m48.217s
```
and after it:
```
real	2m32.053s
user	8m54.281s
sys	0m47.533s

```
so a total gain of around 5%.

On a 64-bit debug build with the sanitizers (`enable-asan enable-ubsan`), before the optimization I got 
```
real	7m43.526s
user	25m32.147s
sys	3m31.876s
```
and after it
```
real	6m53.909s
user	23m21.168s
sys	3m16.811s

```
so a total gain of around 9%.

**Comment by DDvO** [comment]
I leave the job of translating the figures into savings of CO2 emissions per year to someone more politically correct than me :stuck_out_tongue_winking_eye:

**Comment by paulidale** [comment]
ERROR: INSUFFICIENT DATA

**Comment by DDvO** [comment]
> ERROR: INSUFFICIENT DATA

I have insufficient data to interpret this message :wink:

**Comment by richsalz** [comment]
Are there still any tests that generate keys?  I didn't find any recipe named "*keygen*" for example.

**Comment by DDvO** [comment]
> Are there still any tests that generate keys? I didn't find any recipe named "_keygen_" for example.

No worries. The basic tests for RSA and DH key generation and handling stay as they are:
`test_*rsa*`, `test_*dh*`, and `test_evp*` stay as they are. 
These include key generation, in particular `test_gendh` and `test_genrsa`.

File: apps/req.c [hunk 1]
```diff
@@ -696,17 +696,21 @@ int req_main(int argc, char **argv)
 
         EVP_PKEY_CTX_free(genctx);
         genctx = NULL;
+    }
```
Comment (t8m) [inline apps/req.c:699]: Could you please separate changes in this file to a separate commit? With appropriate commit message, of course. It is not clear what this is fixing/changing.

File: test/CAtsa.cnf [hunk 2]
```diff
@@ -48,7 +48,7 @@ emailAddress		= optional
 
 #----------------------------------------------------------------------
 [ req ]
-default_bits		= 2048
+#default_bits		= 2048 # keys are now pre-loaded for efficient tests
```
Comment (t8m) [inline test/CAtsa.cnf:51]: I'd just drop this change. Is it necessary?

File: test/ca-and-certs.cnf [hunk 3]
```diff
@@ -3,7 +3,7 @@ CN2 = Brother 2
 
 ####################################################################
 [ req ]
-default_bits		= 2048
+#default_bits		= 2048 # keys are now pre-loaded for efficient tests
```
Comment (t8m) [inline test/ca-and-certs.cnf:6]: Same here in this file.

File: test/ca-and-certs.cnf [hunk 4]
```diff
@@ -19,7 +19,7 @@ commonName_value		= Dodgy CA
 
 ####################################################################
 [ userreq ]
-default_bits		= 2048
+#default_bits		= 2048 # keys are now pre-loaded for efficient tests
```
Comment (t8m) [inline test/ca-and-certs.cnf:22]: And here.

File: test/proxy.cnf [hunk 5]
```diff
@@ -2,7 +2,7 @@
 ## Config file for proxy certificate testing.
 
 [ req ]
-default_bits		= 2048
+#default_bits		= 2048 # keys are now pre-loaded for efficient tests
```
Comment (t8m) [inline test/proxy.cnf:5]: and here

File: test/test.cnf [hunk 6]
```diff
@@ -49,14 +49,14 @@ emailAddress		= optional
 
 ####################################################################
 [ req ]
-default_bits		= 2048
+default_bits		= 2048 # TODO avoid using this, for more efficient tests
```
Comment (t8m) [inline test/test.cnf:52]: Please drop this. This TODO brings no value.

File: test/test.cnf [hunk 7]
```diff
@@ -49,14 +49,14 @@ emailAddress		= optional
...
 default_keyfile 	= testkey.pem
 distinguished_name	= req_distinguished_name
 encrypt_rsa_key		= no
 
 # Make altreq be identical to req
 [ altreq ]
-default_bits		= 2048
+default_bits		= 2048 # TODO avoid using this, for more efficient tests
```
Comment (t8m) [inline test/test.cnf:59]: And here as well.

**Review by t8m — CHANGES_REQUESTED** [review]
Some nits to cleanup.

File: apps/req.c (hunk 1 above)
Comment (DDvO) [reply apps/req.c:679]: Ah, yes, this should have been more explicit.
The point is that, so far, the `-keyout` option was ignored unless a new key is created.
With the change, which is now in a separate commit, the `keyout` option respected also in case the `-key` option is given.
This is needed by the improved tests that no more use the key generation feature of `req` 
but a pre-computed key from the file now given using the `key` option.
On this occasion I also improved the documentation of the `req` options `-new`, `-newkey`, and `-keyout`.

File: test/CAtsa.cnf (hunk 2 above)
Comment (DDvO) [reply test/CAtsa.cnf]: With the comment I wanted to make clear that the `default_bits` option is no more used,
since otherwise it may be confusing.
Yet in fact I meanwhile think it's even better to entirely remove both the `default_bits` and `default_keyfile` options from those test `.cnf` files because both these options are no more used. Done this now.

File: test/test.cnf (hunk 6 above)
Comment (DDvO) [reply test/test.cnf]: This (and the one mentioned below) was a leftover that I forgot to handle.
Solved as I wrote above.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
> Are there still any tests that generate keys?

For curiosity, made a quick experiment replacing all RSA and DH key lengths in the generation functions by 512.
This further reduced the test run time (here for a 64-bit plain debug build on my machine) from
```
real	2m32.053s
user	8m54.281s
sys	0m47.533s
```
 to
```
real	2m7.137s
user	7m31.989s
sys	0m42.072s
```
So the expensive key generation operations account for some 10% of the total test time,
and for sanitizer builds certainly even more.

**Comment by richsalz** [comment]
> Changing to 512 bit keys.

Do not do this. Do not use weak/small keys in tests just to shorten testing time.

Keys should be 2K ideally.

**Comment by DDvO** [comment]
> > Changing to 512 bit keys.
> 
> Do not do this. Do not use weak/small keys in tests just to shorten testing time.

That was already discussed and agreed on above.

> Keys should be 2K ideally.

Generally yes, except when one wanna test that key generation works fine also for smaller or larger keys.
The current tests for RSA go as far as 8192 bits.

**Comment by DDvO** [comment]
> > That was already discussed and agreed on above.
> 
> I hope the OTC will step in and consider this "policy."

Apparently you misunderstood. 
We won't shorten key lengths just to speed up tests.

**Comment by richsalz** [comment]
Yes I misunderstood, thanks for the clarification @ddvo.

**Review by paulidale — APPROVED** [review]
re-approved

**Comment by DDvO** [comment]
Merged - thanks to all who contributed in the discussion and approval.
